gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+                "s22": s22,
+            }
+        )
+
+        length_sweep2 = pd.DataFrame.from_dict(
+            {
+                "length": [L[0] for L in group_span],
+                "s11": list(abs(S["s11"]) ** 2),
+                "s21": list(abs(S["s21"]) ** 2),
+                "s12": list(abs(S["s12"]) ** 2),
+                "s22": list(abs(S["s22"]) ** 2),
+            }
+        )
+
+        length_sweep2.to_csv(
+            str(self.dirpath / f"{self.component.name}_length_sweep.csv")
+        )
+        return length_sweep
+
+    def plot_length_sweep(
+        self,
+        start_length: float = 1,
+        stop_length: float = 100,
+        num_pts: int = 100,
+        group: int = 1,
+    ) -> None:
+        """
+        Plot length sweep.
+
+        Parameters:
+            start_length: Start length (um)
+            stop_length: Stop length (um)
+            num_pts: Number of points along sweep
+            group: Group span to sweep. First group starts from 0.
+
+        Returns:
+            Figure handle
+        """
+        sweep_data = self.get_length_sweep(start_length, stop_length, num_pts, group)
+
+        fig = plt.figure()
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s11"]) ** 2)
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s21"]) ** 2)
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s12"]) ** 2)
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s22"]) ** 2)
+        plt.legend(["|S11|^2", "|S21|^2", "|S12|^2", "|S22|^2"])
+        plt.grid("on")
+        plt.xlabel("Length (um)")
+        plt.ylabel("Magnitude")
+        plt.title(f"Length Sweep | Wavelength={self.simulation_settings.wavelength}um")
+        plt.tight_layout()
+        plt.savefig(str(self.dirpath / f"{self.component.name}_length_sweep.png"))
+
+        return fig
+
+    def get_mode_coupling(
+        self, input_mode: int, max_coupled_mode: int = None, group: int = 1
+    ) -> dict[str, pd.DataFrame]:
+        """
+        Get mode coupling coefficients from the input_mode to the max_coupled_mode.
+        Mode numbers start from 1 (fundamental mode).
+
+        Parameters:
+            input_mode: Mode of interest. Energy couples away from this mode to other modes.
+            max_coupled_mode: Maximum mode number to consider; all modes below this mode will be considered as well.
+            group: Group of cells to consider. First group is 0.
+
+        Returns:
+            Dictionary of coupling coefficients vs. position along device:
+            1. forward_sparam - Forward sparam coupling (dB)
+            | mode1 | mode2 | ...
+            | float | float | ...
+            2. backward_sparam Backward sparam coupling (dB)
+            | mode1 | mode2 | ...
+            | float | float | ...
+            3. overlap - Overlap integrals (dB)
+            | mode1 | mode2 | ...
+            | float | float | ...
```

Comment:
sure, should still specify that it's not a scalar float

Reply:
It is a column of scalar floats. If each entry has an array, it would be specified like:
https://github.com/elamdf/gplugins/pull/3#discussion_r1518384056

Thoughts on documenting this way?